jina/types/document/__init__.py

```diff
@@ -0,0 +1,194 @@
+        for d in self._document.matches:
+            yield Document(d)
+
+    @property
+    def chunks(self) -> Iterator['Document']:
+        for d in self._document.chunks:
+            yield Document(d)
+
+    def add_match(self, doc_id: str, score_value: float, **kwargs) -> 'Document':
+        r = self._document.matches.add()
+        r.id = doc_id
+        r.granularity = self._document.granularity
+        r.adjacency = self._document.adjacency + 1
+        r.score.ref_id = self._document.id
+        r.score.value = score_value
+        for k, v in kwargs.items():
+            if hasattr(r.score, k):
+                setattr(r.score, k, v)
+        return Document(r)
+
+    def add_chunk(self, document: Optional['Document'] = None, **kwargs) -> 'Document':
+        c = self._document.chunks.add()
+        if document is not None:
+            c.CopyFrom(document.as_pb_object)
+
+        with Document(c) as chunk:
+            chunk.update(parent_id=self._document.id,
+                         granularity=self._document.granularity + 1,
+                         mime_type=self._document.mime_type,
+                         **kwargs)
+            return chunk
+
+    def update(self, **kwargs):
+        """Bulk update Document fields with key-value specified in kwargs """
+        for k, v in kwargs.items():
+            if isinstance(v, list) or isinstance(v, tuple):
+                self._document.ClearField(k)
+                getattr(self._document, k).extend(v)
+            elif isinstance(v, dict):
+                self._document.ClearField(k)
+                getattr(self._document, k).update(v)
+            else:
+                setattr(self, k, v)
+
+    @cached_property
+    def as_pb_object(self) -> 'jina_pb2.DocumentProto':
+        return self._document
+
+    @property
+    def buffer(self) -> bytes:
+        return self._document.buffer
+
+    @buffer.setter
+    def buffer(self, value: bytes):
+        self._document.buffer = value
+        if not self._document.mime_type and _buffer_sniff:
+            try:
+                import magic
+                self._document.mime_type = magic.from_buffer(value, mime=True)
+            except Exception as ex:
+                default_logger.warning(f'can not sniff the MIME type: {repr(ex)}')
+
+    @property
+    def uri(self) -> str:
+        return self._document.uri
+
+    @uri.setter
+    def uri(self, value: str):
+        scheme = urllib.parse.urlparse(value).scheme
+        if ((scheme in {'http', 'https'} and is_url(value))
+                or (scheme in {'data'})
+                or os.path.exists(value)
+                or os.access(os.path.dirname(value), os.W_OK)):
+            self._document.uri = value
+            self._document.mime_type = guess_mime(value)
+        else:
+            raise ValueError(f'{value} is not a valid URI')
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.update_id()
```

Comment:
I think this is a little dangerous, people may uncounciously update the `id` of a document which is a very sensitive thing to do. I would leave it as something to be done explicitly.

Reply:
will expose an interface. this is nothing near final version. I will adapt all interfaces according to our real usage in the code.